src/coreclr/jit/inlinepolicy.cpp

```diff
@@ -1440,6 +1440,71 @@ void ExtendedDefaultPolicy::NoteDouble(InlineObservation obs, double value)
     m_ProfileFrequency = value;
 }
 
+//------------------------------------------------------------------------
+// BudgetCheck: see if this inline would exceed the current budget
+//
+// Returns:
+//   True if inline would exceed the budget.
+//
+bool ExtendedDefaultPolicy::BudgetCheck() const
+{
+    if (m_IsPrejitRoot)
+    {
+        // Only relevant if we're actually inlining.
+        return false;
+    }
+
+    // The strategy tracks the amout of inlining done so far, so it performs the actual check.
+    const bool overBudget = m_RootCompiler->m_inlineStrategy->BudgetCheck(m_CodeSize);
+
+    if (!overBudget)
+    {
+        // We're not over budget
+        return false;
+    }
+
+    assert(m_IsForceInlineKnown);
+    if (!m_IsForceInline)
+    {
+        // For over-budget case we only inspect forceinline
+        // Otherwise, report "over-budget"
+        return true;
+    }
+
+    if (m_CallsiteDepth == 1)
```

Comment:
Is there any way we can better track `m_CallsiteDepth` for trivial cases?

That is, there are many scenarios where you might have a code pattern like the following:
```
public void PublicApi(...)
{
    // ArgumentValidation
    InternalApi(...);
}

void InternalApi(...)
{
    if (Vector256.IsHardwareAccelerated)
    {
        InternalApiV256();
    }
    else if (Vector128.IsHardwareAccelerated)
    {
        InternalApiV128();
    }
    else
    {
        InternalApiScalar();
    }
}
```

Such scenarios involve a couple small almost "stub" methods or methods that just forward to another (another example is `int.LeadingZeroCount` which just forwards to `BitOperations.LeadingZeroCount`).

In such scenarios, we will almost always by at `m_CallsiteDepth >= 2` and not have gotten to the actual interesting "core" yet.

If we had a way to factor in such methods that are effectively just "forward" or "validate + forward" then we could more meaningfully do inlining for such cases.

Reply:
I've changed the current impl, it now tries to guess final imported IL size. so for your case InternalApi won't affect the time budget since the majority of code will be folded anyway. So e.g. InternalApiV256 will have plenty of time budget left to be inlined (especially if inliner can detect foldable branches in it).